dmoj/cptbox/isolate.py

```diff
@@ -314,63 +348,97 @@ def _file_access_check(
         # This works, except when the child process uses /proc/self, which refers to something else in this process.
         # Therefore, we "project" it by changing it to /proc/[tid] for computing the realpath and doing the samefile
         # check. However, we still keep it as /proc/self when checking access rules.
+
+        # normpath doesn't strip leading slashes
         projected = normalized = '/' + os.path.normpath(file).lstrip('/')
+        proc_dir = f'/proc/{debugger.tid}'
         if normalized.startswith('/proc/self'):
-            file = os.path.join(f'/proc/{debugger.tid}', os.path.relpath(file, '/proc/self'))
-            projected = '/' + os.path.normpath(file).lstrip('/')
-        elif normalized.startswith(f'/proc/{debugger.tid}/'):
+            projected = os.path.join(proc_dir, os.path.relpath(projected, '/proc/self'))
+        elif normalized.startswith(proc_dir):
             # If the child process uses /proc/getpid()/foo, set the normalized path to be /proc/self/foo.
             # Access rules can more easily check /proc/self.
-            normalized = os.path.join('/proc/self', os.path.relpath(file, f'/proc/{debugger.tid}'))
-        real = os.path.realpath(file)
+            normalized = os.path.join('/proc/self', os.path.relpath(normalized, proc_dir))
 
+        real = os.path.realpath(projected)
         try:
-            same = normalized == real or os.path.samefile(projected, real)
+            same = projected == real or os.path.samefile(projected, real)
         except OSError:
-            log.debug('Denying access due to inability to stat: normalizes to: %s, actually: %s', normalized, real)
-            return file, ACCESS_ENOENT
-        else:
-            if not same:
-                log.warning(
-                    'Denying access due to suspected symlink trickery: normalizes to: %s, actually: %s',
-                    normalized,
-                    real,
-                )
-                return file, ACCESS_EACCES
+            raise AccessFailure(ACCESS_ENOENT, 'Cannot stat, file: %s, projected: %s, real: %s', file, projected, real)
```

Comment:
```suggestion
            raise AccessFailure(ACCESS_ENOENT, f'Cannot stat, {file=}, {projected=}, {real=}')
```

Reply:
Isn't it better to not use an f-string? Since these are all being passed to a `log` function, which might not need to output the string, it's better not to use CPU cycles, right?

Reply:
The formatting is a drop in the bucket to everything else. If we cared about performance of trapped syscalls, we wouldn't be doing things in allocating Python code.